mava/systems/q_learning/anakin/rec_qmix.py

```diff
@@ -0,0 +1,697 @@
+        if cfg.system.hard_update:
+            next_target_params = optax.periodic_update(
+                next_online_params, params.target, t_train, cfg.system.update_period
+            )
+            next_mixer_target_params = optax.periodic_update(
+                next_mixer_params, params.mixer_target, t_train, cfg.system.update_period
+            )
+        else:
+            next_target_params = optax.incremental_update(
+                next_online_params, params.target, cfg.system.tau
+            )
+            next_mixer_target_params = optax.incremental_update(
+                next_mixer_params, params.mixer_target, cfg.system.tau
+            )
+        # Repack params and opt_states.
+        next_params = QMIXParams(
+            next_online_params,
+            next_target_params,
+            next_mixer_params,
+            next_mixer_target_params,
+        )
+
+        return next_params, next_opt_state, q_loss_info
+
+    def train(
+        train_state: TrainState[QMIXParams], _: Any
+    ) -> Tuple[TrainState[QMIXParams], Metrics]:
+        """Sample, train and repack."""
+
+        buffer_state, params, opt_states, t_train, key = train_state
+        next_key, buff_key = jax.random.split(key, 2)
+
+        data = rb.sample(buffer_state, buff_key).experience
+
+        # Learn
+        next_params, next_opt_states, q_loss_info = update_q(params, opt_states, data, t_train)
+
+        next_train_state = TrainState(
+            buffer_state, next_params, next_opt_states, t_train + 1, next_key
+        )
+
+        return next_train_state, q_loss_info
+
+    # ---- Act-train loop ----
+    scanned_act = lambda state: lax.scan(action_step, state, None, length=cfg.system.rollout_length)
+    scanned_train = lambda state: lax.scan(train, state, None, length=cfg.system.epochs)
+
+    # Act and train
+    def update_step(
+        learner_state: LearnerState, _: Any
+    ) -> Tuple[LearnerState, Tuple[Metrics, Metrics]]:
+        """Act, then learn."""
+
+        (
+            obs,
+            terminal,
+            term_or_trunc,
+            hidden_state,
+            env_state,
+            time_steps,
+            train_steps,
+            opt_state,
+            buffer_state,
+            params,
+            key,
+        ) = learner_state
+        new_key, act_key, train_key = jax.random.split(key, 3)
+
+        # Select actions, step env and store transitions
+        action_selection_state = ActionSelectionState(
+            params.online, hidden_state, time_steps, act_key
+        )
+        action_state = ActionState(
+            action_selection_state, env_state, buffer_state, obs, terminal, term_or_trunc
+        )
+        final_action_state, metrics = scanned_act(action_state)
+
+        # Sample and learn
+        train_state = TrainState(
+            final_action_state.buffer_state,
+            params,
+            opt_state,
+            train_steps,
+            train_key,
+        )
```

Comment:
nit formatting to align with IQL
```suggestion
        train_state = TrainState(
            final_action_state.buffer_state, params, opt_state, train_steps, train_key
        )
```